packages/node-sdk/sm/primitives/evm/rpc/paima-l2.ts

```diff
@@ -1,185 +1,169 @@
-    // guarantee we run this no matter if there is an error or a continue
-    await insertNonce.run(
-      {
-        nonce: inputData.inputNonce,
-        block_height: latestChainData.blockNumber,
-      },
-      DBConn,
+  } catch (err) {
+    log.remote(
+      ComponentNames.PAIMA_SYNC,
+      ["paima-l2"],
+      SeverityNumber.ERROR,
+      (log) => log(`[paima-sm] Error on user input STF call. Skipping`, err),
     );
-    if (ENV.STORE_HISTORICAL_GAME_INPUTS) {
-      await newGameInput.run(
-        {
-          block_height: latestChainData.blockNumber,
-          input_data: inputData.inputData,
-          from_address: inputData.userAddress,
-          success: success ?? false,
-          paima_tx_hash: Buffer.from(txHash, "hex"),
-          index_in_block: txIndexInBlock,
-          origin_tx_hash: Buffer.from(strip0x(inputData.origin.txHash!), "hex"),
-          caip2: inputData.origin.caip2!,
-          primitive_name: inputData.origin.primitiveName ?? "",
-          origin_contract_address: inputData.origin.contractAddress,
-        },
-        DBConn,
-      );
-    }
-    txIndexInBlock += 1;
   }
 
-  const primitiveName = response.output.syncProtocol.payload.primitiveName;
-  const { from, value } = response.output.payload;
-
-  const fromAddr = from.toLowerCase() as EvmAddress;
-
-  const fromRow = yield* World.resolve(
-    primitiveErc20DepositGetTotalDeposited,
-    {
-      primitive_name: primitiveName,
-      wallet_address: fromAddr,
-    },
+  yield* StateMachineExecution(
+    input.paima_block_height,
+    hexToString((input.payload as any).data),
+    address.address as `0x${string}`,
+    address.id,
+    input.ownChain.blockNumber,
+    input.ownChain.transactionHash,
   );
+}
 
-  const numValue = BigInt(value);
-  const prefix = response.input.scheduledPrefix;
+export default function* processPaimaL2SyncProtocolResponse(
+  paima_block_height: BlockNumber,
+  response: FlattenSyncProtocolIOFor<
+    | ConfigSyncProtocolType.EVM_RPC_MAIN
+    | ConfigSyncProtocolType.EVM_RPC_PARALLEL,
+    ConfigPrimitiveType.EvmRpcPaimaL2,
+    ConfigPrimitivePayloadType.PaimaL2Event
+  >,
+): StateUpdateStream<void> {
+  const outerLayerPayload = clearBigInts(response.output.payload);
 
+  let isBatched = false;
   try {
-    const scheduledInputData = generateRawStmInput(
-      BuiltinTransitions[ConfigPrimitiveType.ERC20Deposit].scheduledPrefix,
-      prefix,
-      {
-        from: fromAddr,
-        value,
-      },
-    );
-    const scheduledBlockHeight =
-      response.output.syncProtocol.payload.mainchain.blockNumber;
-    yield* createScheduledData(
-      JSON.stringify(scheduledInputData),
-      { blockHeight: scheduledBlockHeight },
-      {
+    const message = hexToString((outerLayerPayload as any).data);
+    const batchedMessages = extractBatches(message);
+    isBatched = true;
+    for (const batchedMessage of batchedMessages) {
+      // TODO we need to verify the signature of the batched messages
```

Comment:
dangerous TODO (but I think this code will all change later, so maybe it's not as important)

Reply:
PAI-26